comb_spec_searcher/equiv_db.py

```diff
@@ -70,16 +71,60 @@ def __getitem__(self, comb_class: int) -> int:
             self.parents[ancestor] = root
         return root
 
-    def add_edge(self, label: int, other_label: int) -> None:
+    def get_one_way_vertices(self) -> Dict[int, Set[int]]:
+        res: Dict[int, Set[int]] = defaultdict(set)
+        for start, ends in self._one_way_vertices.items():
+            start = self[start]
+            for end in ends:
+                end = self[end]
+                if start != end:
+                    res[start].add(end)
+        self._one_way_vertices = res
+        return res
+
+    def add_one_way_edge(self, label: int, other_label: int) -> None:
+        self._add_edge(label, other_label)
+        self._one_way_vertices[self[label]].add(self[other_label])
+
+    @cssmethodtimer("find_paths")
+    def connect_cycles(self):
+        """Look for cycles using one way edges that have been added."""
+        one_way_vertices = self.get_one_way_vertices()
+        dequeue: Deque[Tuple[int, ...]] = deque()
+        dequeue.extend([(label,) for label in one_way_vertices.keys()])
+        visited = set()
+        while dequeue:
+            path = dequeue.popleft()
+            end = path[-1]
+            if end in visited:
```

Comment:
I don't think you can continue here. Not 100% sure but since you have many paths starting from different vertices my intuition says it doesn't work

Reply:
We are searching depth first. If end is in visited, then there is already a path on the queue which considers the edges coming out of end.

Reply:
Not on the queue, because it is depth first, it means we have already checked if there is a cycle using end.

Reply:
I'm a little confused here. I don't see how this is depth first search since you are actually traversing using a queue

Reply:
Ok I actually notice that on line 113 you append left to dequeue. So you use it like a stack. This confusing should probably change the variable name to stack and use a normal list with append and pop

Reply:
yeah we could change that